en/news/_posts/2019-12-06-separation-of-positional-and-keyword-arguments-in-ruby-3-0.md

```diff
@@ -0,0 +1,377 @@
+
+## tl;dr
+
+In Ruby 3.0, positional arguments and keyword arguments will be separated.  Ruby 2.7 will warn for behavior that will change in Ruby 3.0.  If you see the following warnings, you will need to update your code:
+
+* `The last argument is used as the keyword parameter`, or
+* `The keyword argument is passed as the last hash parameter`, or
+* `The last argument is split into positional and keyword parameters`
+
+Typically, you can fix the issue by using the "double splat" operator to pass keywords instead of a Hash object, or writing braces to pass a Hash object instead of keywords.  See the next section.
+
+A method that *delegates* arguments needs to delegate keyword arguments explicitly in Ruby 3.  If you want to write compatible delegation code that works on older Ruby versions and Ruby 3, use `ruby2_keywords`.  See the "Handling argument delegation" section in detail.
+
+## Typical cases
+
+Here is the most typical case. You can use double splat operator (`**`) to pass keywords instead of a Hash.
+
+```
+# This method accepts only a keyword argument
+def foo(k: 1)
+  p k
+end
+
+h = { k: 42 }
+
+# This method call passes a positional Hash argument
+# In Ruby 2.7: The Hash is automatically converted to a keyword argument
+# In Ruby 3.0: This call raises an ArgumentError
+foo(h)
+  # => demo.rb:11: warning: The last argument is used as the keyword parameter
+  #    demo.rb:2: warning: for `foo' defined here; maybe ** should be added to the call?
+  #    42
+
+# If you want to keep the behavior in Ruby 3.0, use double splat
+foo(**h) #=> 42
+```
+
+Here is another case.  You can use braces to pass a Hash instead of keywords explicitly.
+
+```
+# This method accepts one positional argument and a keyword rest argument
+def bar(h, **kwargs)
+  p h
+end
+
+# This call passes only a keyword argument and no positional arguments
+# In Ruby 2.7: The keyword is converted to a positional Hash argument
+# In Ruby 3.0: This call raises an ArgumentError
+bar(k: 42)
+  # => demo2.rb:9: warning: The keyword argument is passed as the last hash parameter
+  #    demo2.rb:2: warning: for `bar' defined here
+  #    {:k=>42}
+
+# If you want to keep the behavior in Ruby 3.0, write braces to make it an
+# explicit Hash
+bar({ k: 42 }) # => {:k=>42}
+```
+
+## What was deprecated?
+
+In Ruby 2, keyword arguments can be treated as the last positional Hash argument and a last positional Hash argument can be treated as keyword arguments.
+
+Because the automatic conversion was sometimes too complex and troublesome (described in the last section), it is deprecated in Ruby 2.7 and will be removed in Ruby 3.  In other words, keyword arguments will be completely separated from positional ones.  If you want to pass keyword arguments, you should always use `foo(k: expr)` or `foo(**expr)`.  If you want to accept keyword arguments, in principle you should always use `def foo(k: value)` or `def foo(k:)` or `def foo(**kw)`.
```

Comment:
What is "(described in the last section)?" I don't think the last section "Typical cases" explaining the ambiguous cases clearly.

Reply:
I meant the final section "Why we're deprecating the automatic conversion", not the previous one.  Is "described in the final section" better?

Reply:
Yes, you definitely want to use `final section` instead of `last section`, as `last section` implies the section you just read, not the section at the end of the document.

Reply:
Fixed, thank you!